src/ILCompiler.Compiler/src/Compiler/PreInitFieldInfo.cs

```diff
@@ -8,9 +8,71 @@
 
 using Internal.TypeSystem;
 using Internal.TypeSystem.Ecma;
+using ILCompiler.DependencyAnalysis;
 
 namespace ILCompiler
 {
+    public abstract class PreInitFixupInfo : IComparable<PreInitFixupInfo>
+    {
+        /// <summary>
+        /// Offset into the blob
+        /// </summary>
+        public int Offset { get; }
+
+        public PreInitFixupInfo(int offset)
+        {
+            Offset = offset;
+        }
+
+        int IComparable<PreInitFixupInfo>.CompareTo(PreInitFixupInfo other)
+        {
+            return this.Offset - other.Offset;
+        }
+
+        /// <summary>
+        /// Writes fixup data into current ObjectDataBuilder. Caller needs to make sure ObjectDataBuilder is
+        /// at correct offset before writing.
+        /// </summary>
+        /// <returns>Bytes written</returns>
+        public abstract int WriteData(ref ObjectDataBuilder builder, NodeFactory factory);
+    }
+
+    public class PreInitTypeFixupInfo : PreInitFixupInfo
+    {
+        public TypeDesc TypeFixup { get; }
+
+        public PreInitTypeFixupInfo(int offset, TypeDesc type)
+            :base(offset)
+        {
+            TypeFixup = type;
+        }
+
+        public override int WriteData(ref ObjectDataBuilder builder, NodeFactory factory)
+        {
+            builder.EmitPointerReloc(factory.NecessaryTypeSymbol(TypeFixup));
+
+            return IntPtr.Size;
+        }
+    }
+
+    public class PreInitMethodFixupInfo : PreInitFixupInfo
+    {
+        public MethodDesc MethodFixup { get; }
+
+        public PreInitMethodFixupInfo(int offset, MethodDesc method)
+            : base(offset)
+        {
+            MethodFixup = method;
+        }
+
+        public override int WriteData(ref ObjectDataBuilder builder, NodeFactory factory)
+        {
+            builder.EmitPointerReloc(factory.MethodEntrypoint(MethodFixup));
```

Comment:
The behavior for `MethodEntrypoint` when called for e.g. `Foo<object>.Method` is to assert the compiler.

We might want to do _one of_ these:
* Use `CanonicalEntrypoint` (this will emit a reloc to `Foo<__Canon>.Method`)
* Use `ExactCallableAddress` (this will emit a fat function pointer that represents `Foo<object>.Method`)
* Assert that `method` is already in the canonical form in the constructor of `PreInitMethodFixupInfo` or make it throw BadImageFormat.